From 57d4ac1722004885c51c9a0c64d1be90a0aa632d Mon Sep 17 00:00:00 2001
From: Haiying Wang <Haiying.Wang@nxp.com>
Date: Thu, 20 Apr 2017 11:54:22 -0400
Subject: [PATCH 570/741] soc: fsl: dpio: change CENA regs to be cacheable

commit 6255bb24e9868ec72f335c6c00ec4002f128659a from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Change cache enabled regsiter accessed to be cacheable
plus non-shareable to meet the performance requirement.
QMan's CENA region contains registers and structures that
are 64byte in size and are inteneded to be accessed using a
single 64 byte bus transaction, therefore this portal
memory should be configured as cache-enabled. Also because
the write allocate stash transcations of QBMan should be
issued as cachable and non-coherent(non-sharable), we
need to configure this region to be non-shareable.

Signed-off-by: Haiying Wang <Haiying.Wang@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/soc/fsl/dpio/dpio-driver.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/soc/fsl/dpio/dpio-driver.c b/drivers/soc/fsl/dpio/dpio-driver.c
index ecfe2ac86f22..18c6d434f8cb 100644
--- a/drivers/soc/fsl/dpio/dpio-driver.c
+++ b/drivers/soc/fsl/dpio/dpio-driver.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0+ OR BSD-3-Clause)
 /*
  * Copyright 2014-2016 Freescale Semiconductor Inc.
- * Copyright 2016-2018 NXP
+ * Copyright 2016-2019 NXP
  *
  */
 
@@ -27,6 +27,9 @@ MODULE_LICENSE("Dual BSD/GPL");
 MODULE_AUTHOR("Freescale Semiconductor, Inc");
 MODULE_DESCRIPTION("DPIO Driver");
 
+#define PROT_NORMAL_NS         (PTE_TYPE_PAGE | PTE_AF | PTE_PXN | PTE_UXN | PTE_DIRTY | PTE_WRITE | PTE_ATTRINDX(MT_NORMAL))
+#define ioremap_cache_ns(addr, size)   __ioremap((addr), (size), __pgprot(PROT_NORMAL_NS))
+
 struct dpio_priv {
 	struct dpaa2_io *io;
 };
@@ -199,6 +202,10 @@ static int dpaa2_dpio_probe(struct fsl_mc_device *dpio_dev)
 
         if (dpio_dev->obj_desc.region_count < 3) {
                 /* No support for DDR backed portals, use classic mapping */
+                /*
+                 * Set the CENA regs to be the cache enabled area of the portal to
+                 * achieve the best performance.
+                 */
                 desc.regs_cena = ioremap_cache_ns(dpio_dev->regions[0].start,
                                         resource_size(&dpio_dev->regions[0]));
         } else {
-- 
2.17.1

